### PR TITLE
Convert end-to-end tests to Typescript

### PR DESCRIPTION
**Pull request by novocaine**
Convert e2e tests to typescript and get their CI to pass. I've uploaded a new docker image to dockerhub based on the changes to the Dockerfile so you can see the green tick here.

This means the tests are compiled as part of `run.sh` and the output `.js` is in `dist`.

I need someone's expertise to help make the Window references work nicely.

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

Preview: https://61ad4208d6fb4c1d7403c414--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: test/end-to-end-tests/run.sh [hunk 1]
```diff
@@ -35,5 +35,6 @@ trap 'handle_error' ERR
 if [ $has_custom_app -ne "1" ]; then
     ./element/start.sh
 fi
-node start.js $@
+yarn build
```
Comment (novocaine) [inline test/end-to-end-tests/run.sh:38]: I've added a build step to run.sh - this seems reasonable as it only takes a few seconds on my machine, but maybe we'll want separate control over it at some point

File: test/end-to-end-tests/src/@types/global.d.ts [hunk 2]
```diff
@@ -15,16 +14,9 @@ See the License for the specific language governing permissions and
...
-        this.buffer = initialValue;
-        page.on(eventName, (arg) => {
-            const result = eventMapper(arg);
-            if (reduceAsync) {
-                result.then((r) => this.buffer += r);
-            } else {
-                this.buffer += result;
-            }
-        });
+declare global {
```
Comment (novocaine) [inline test/end-to-end-tests/src/@types/global.d.ts:17]: Would appreciate some help with this as it doesn't quite work - references to `window.mxPerformanceMonitor` still trigger type errors, which I'm working around by calling it `window['mxPerformanceMonitor']`

File: test/end-to-end-tests/start.ts [hunk 3]
```diff
@@ -33,6 +33,14 @@ program
 
 const hsUrl = 'http://localhost:5005';
 
+// XXX: why is this not located in @types/global.d.ts?
```
Comment (novocaine) [inline test/end-to-end-tests/start.ts:36]: references to window.mxPerformanceMonitor in this file don't seem to be using the definition from `@types/global.d.ts`

File: test/end-to-end-tests/src/@types/global.d.ts (hunk 2 above)
Comment (turt2live) [reply test/end-to-end-tests/src/@types/global.d.ts:17]: I'm a bit surprised to see e2e-test specific declarations tbh. In any case, the types will need to be imported from somewhere central (or each test if not central) - I'd recommend importing the real types from the react-sdk src instead.

If importing the real ones causes issues (like import loops), then comments on the `any` usage would be preferred.

File: test/end-to-end-tests/src/logger.ts [hunk 4]
```diff
@@ -15,14 +15,13 @@ See the License for the specific language governing permissions and
...
 
-module.exports = class Logger {
-    constructor(username) {
-        this.indent = 0;
-        this.username = username;
-        this.muted = false;
-    }
+export class Logger {
+    indent = 0;
+    muted = false;
```
Comment (t3chguy) [inline test/end-to-end-tests/src/logger.ts:20]: visibilities please

File: test/end-to-end-tests/src/rest/consent.ts [hunk 5]
```diff
@@ -15,11 +15,11 @@ See the License for the specific language governing permissions and
...
 
-const request = require('request-promise-native');
-const cheerio = require('cheerio');
-const url = require("url");
+import request = require('request-promise-native');
+import * as cheerio from 'cheerio';
+import * as url from "url";
 
-module.exports.approveConsent = async function(consentUrl) {
+export const approveConsent = async function(consentUrl: string): Promise<any> {
```
Comment (t3chguy) [inline test/end-to-end-tests/src/rest/consent.ts:22]: why any?

File: test/end-to-end-tests/src/rest/creator.ts [hunk 6]
```diff
@@ -32,27 +37,31 @@ function execAsync(command, options) {
...
+    userId: string;
+    deviceId: string;
+    hsUrl: string;
+}
+
+export class RestSessionCreator {
+    constructor(readonly synapseSubdir: string, readonly hsUrl: string, readonly cwd: string) {}
 
-    async createSessionRange(usernames, password, groupName) {
+    async createSessionRange(usernames: string[], password: string, groupName: string): Promise<RestMultiSession> {
```
Comment (t3chguy) [inline test/end-to-end-tests/src/rest/creator.ts:51]: visibilities on async methods please

File: test/end-to-end-tests/src/rest/multi.ts [hunk 7]
```diff
@@ -15,19 +15,22 @@ See the License for the specific language governing permissions and
...
+import { Logger } from '../logger';
+import { RestSession } from "./session";
+import { RestRoom } from "./room";
 
-module.exports = class RestMultiSession {
-    constructor(sessions, groupName) {
+export class RestMultiSession {
+    readonly log: Logger;
+
+    constructor(readonly sessions: RestSession[], groupName: string) {
```
Comment (t3chguy) [inline test/end-to-end-tests/src/rest/multi.ts:25]: visibilities on methods in this class please

File: test/end-to-end-tests/src/rest/multi.ts [hunk 8]
```diff
@@ -15,19 +15,22 @@ See the License for the specific language governing permissions and
...
 
-const Logger = require('../logger');
+import { Logger } from '../logger';
+import { RestSession } from "./session";
+import { RestRoom } from "./room";
 
-module.exports = class RestMultiSession {
-    constructor(sessions, groupName) {
+export class RestMultiSession {
+    readonly log: Logger;
```
Comment (t3chguy) [inline test/end-to-end-tests/src/rest/multi.ts:23]: visibility?

File: test/end-to-end-tests/src/rest/multi.ts [hunk 9]
```diff
@@ -36,7 +39,7 @@ module.exports = class RestMultiSession {
         return session;
     }
 
-    async setDisplayName(fn) {
+    async setDisplayName(fn: (string) => string): Promise<void> {
```
Comment (t3chguy) [inline test/end-to-end-tests/src/rest/multi.ts:42]:
```suggestion
    async setDisplayName(fn: (s: string) => string): Promise<void> {
```

File: test/end-to-end-tests/src/rest/session.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    invite: string;
+    public: boolean;
+    topic: string;
+    dm: boolean;
+}
+
+export class RestSession {
+    private _displayName: string = null;
+    private readonly rooms: Record<string, RestRoom> = {};
+    readonly log: Logger;
```
Comment (t3chguy) [inline test/end-to-end-tests/src/rest/session.ts:33]: visibility?

File: test/end-to-end-tests/src/rest/session.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+    constructor(private readonly credentials: Credentials) {
```
Comment (t3chguy) [inline test/end-to-end-tests/src/rest/session.ts:35]: Can we specify method visibility on all methods in this class?

File: test/end-to-end-tests/src/rest/session.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+    post(csApiPath: string, body?: any): Promise<any> {
+        return this.request("POST", csApiPath, body);
+    }
+
+    put(csApiPath: string, body?: any): Promise<any> {
+        return this.request("PUT", csApiPath, body);
+    }
+
+    async request(method: string, csApiPath: string, body?: any): Promise<any> {
```
Comment (t3chguy) [inline test/end-to-end-tests/src/rest/session.ts:111]: Can we do better here, maybe `Record<string, any>` at the very least

File: test/end-to-end-tests/src/scenario.ts [hunk 13]
```diff
@@ -14,15 +14,19 @@ See the License for the specific language governing permissions and
...
+import { toastScenarios } from './scenarios/toast';
+import { roomDirectoryScenarios } from './scenarios/directory';
+import { lazyLoadingScenarios } from './scenarios/lazy-loading';
+import { e2eEncryptionScenarios } from './scenarios/e2e-encryption';
+import { ElementSession } from "./session";
+import { RestSessionCreator } from "./rest/creator";
+import { RestMultiSession } from "./rest/multi";
+import { spacesScenarios } from './scenarios/spaces';
+
+export async function scenario(createSession: (string) => Promise<ElementSession>,
```
Comment (t3chguy) [inline test/end-to-end-tests/src/scenario.ts:28]:
```suggestion
export async function scenario(createSession: (s: string) => Promise<ElementSession>,
```

File: test/end-to-end-tests/start.ts (hunk 3 above)
Comment (t3chguy) [reply test/end-to-end-tests/start.ts:36]: I think creating a `@types/global.d.ts` relative to the src directory with an import of the react-sdk global.d.ts should do the trick, this is what element-web does.

File: test/end-to-end-tests/src/rest/session.ts (hunk 12 above)
Comment (novocaine) [reply test/end-to-end-tests/src/rest/session.ts:111]: well its a JSON response, so it could be a scalar or `[]` .. effectively, `any`

Comment (t3chguy) [reply test/end-to-end-tests/src/rest/session.ts:111]: I don't think any Matrix APIs return `[]` or `null` though

File: test/end-to-end-tests/src/logbuffer.ts [hunk 14]
```diff
@@ -15,16 +15,28 @@ See the License for the specific language governing permissions and
...
+    return reduceAsync;
+}
+
+export class LogBuffer {
+    buffer: string;
+
+    constructor(page: Page, eventName: keyof PageEventObject, eventMapper: EventMapperSync | EventMapperAsync,
+        reduceAsync = false, initialValue = "") {
         this.buffer = initialValue;
         page.on(eventName, (arg) => {
```
[the remark this answers is not in the document]
Comment (novocaine) [reply test/end-to-end-tests/src/logbuffer.ts:39]: hrm, is there a place in the codebase where we declare types for args with default values? 

webstorm actually deletes the types if I put them in, and it seems typescript infers the type from the default according to https://www.typescriptlang.org/docs/handbook/2/functions.html

File: test/end-to-end-tests/src/rest/session.ts (hunk 12 above)
Comment (novocaine) [reply test/end-to-end-tests/src/rest/session.ts:111]: the types should reflect the deterministic reality .. some matrix API could return that in future, or the server could choose to send back plaintext due to a config error, etc etc

until we actually validate the value, its an `any`

File: test/end-to-end-tests/src/rest/consent.ts (hunk 5 above)
Comment (novocaine) [reply test/end-to-end-tests/src/rest/consent.ts]: oops

File: test/end-to-end-tests/src/rest/multi.ts (hunk 9 above)
Comment (novocaine) [reply test/end-to-end-tests/src/rest/multi.ts]: thanks, this was actually wrong

File: test/end-to-end-tests/src/rest/creator.ts [hunk 15]
```diff
@@ -32,27 +37,32 @@ function execAsync(command, options) {
...
+export interface Credentials {
+    accessToken: string;
+    homeServer: string;
+    userId: string;
+    deviceId: string;
+    hsUrl: string;
+}
+
+export class RestSessionCreator {
+    constructor(readonly synapseSubdir: string, readonly hsUrl: string, readonly cwd: string) {}
```
Comment (t3chguy) [inline test/end-to-end-tests/src/rest/creator.ts:49]: visibilities on the inlined members?

File: test/end-to-end-tests/src/rest/multi.ts [hunk 16]
```diff
@@ -58,30 +61,26 @@ module.exports = class RestMultiSession {
...
-};
+}
 
 class RestMultiRoom {
-    constructor(rooms, roomIdOrAlias, log) {
-        this.rooms = rooms;
-        this.roomIdOrAlias = roomIdOrAlias;
-        this.log = log;
-    }
+    constructor(readonly rooms: RestRoom[], readonly roomIdOrAlias: string, readonly log: Logger) {}
```
Comment (t3chguy) [inline test/end-to-end-tests/src/rest/multi.ts:71]: visibility on these inlined members?

File: test/end-to-end-tests/package.json [hunk 17]
```diff
@@ -4,11 +4,13 @@
...
     "main": "index.js",
     "scripts": {
-        "test": "echo \"Error: no test specified\" && exit 1"
+        "test": "echo \"Error: no test specified\" && exit 1",
+        "build": "tsc -p ./tsconfig.json"
     },
     "author": "",
     "license": "ISC",
     "dependencies": {
+        "@types/puppeteer": "^5.4.4",
```
Comment (t3chguy) [inline test/end-to-end-tests/package.json:13]: should this not be in devDeps?

**Review by t3chguy — APPROVED** [review]
Looks sane to me :shipit:

**Pull request merged**